### DISCUSSION TITLE
Non-maximized windows don't maintain their relative position when moved to the previous/next display

### DISCUSSION THREAD

**Discussion by tahmed42** [rxhanson/Rectangle#781] [General]
- [X] This is a new bug that hasn't been documented in a closed issue or in the Readme.

macOS version: 12.2.1
Rectangle version: 0.53 (59)
Logs if applicable (In Rectangle menu, hold option, "View Logging..."): 

I haven't found anything in this repo yet that states this is happening by purposeful design, so I'm posting this a bug report, but welcome to hear otherwise. Aside from that, it's also possible that I'm hitting an edge case with multiple portrait monitors.

3 4K Dell P2415Q displays, each in portrait (2160 x 3840) scaled to 1440 x 2560 each in macOS.

**Screenshot 1:** At first, there's a TextEdit window in the middle display.
![rectangle_screenshot_1](https://user-images.githubusercontent.com/649696/158261309-c76740aa-25eb-4057-9746-3e8f6f593a95.png)

**Screenshot 2:** Then, we hotkey with Rectangle to make the window snap to the top 50% of the middle display.
![rectangle_screenshot_2](https://user-images.githubusercontent.com/649696/158261337-e6ea10a6-8fec-4c3f-a5b3-f6b6845355b1.png)

**Screenshot 3: Then we hotkey with Rectangle to send the window to the next display. This results in the window moving to the next display, however it also now shifts to the vertical center of the display. I would intuit/expect (and has been the case with ShiftIt), that the window would instead slide over to occupy the same top 50% area that it occupied on the prior display but just on the new display.**
![rectangle_screenshot_3](https://user-images.githubusercontent.com/649696/158261371-f0238f6e-2fe2-4471-a35d-bef192c94e3c.png)

**Screenshot 4:** Next, we hotkey with Rectangle to maximize the window on the current display.
![rectangle_screenshot_4](https://user-images.githubusercontent.com/649696/158261385-548ece77-9ecc-45b9-827e-f230b83671fc.png)

**Screenshot 5:** Next, we hotkey with Rectangle to send the window to the previous display. This results in the window moving to the previous display, and it also nicely maintains its maximized status on the previous display (as hoped for earlier in Screenshot 3).
![rectangle_screenshot_5](https://user-images.githubusercontent.com/649696/158261398-1ab16484-412e-44f4-8928-69c86e4cb859.png)

**Comment by rxhanson** [comment]
Thanks for the detailed report with screenshots. There is some ambiguity here, and the implementation in Rectangle favors simplicity and disregards assumptions on size/placement, mainly considering that displays are not all identical size. It would make sense to add in some logic that preserves which screen edges a window is on, but that would also have to take into account gaps if configured. 

There is a hidden command that I haven't published, that you can use
`defaults write com.knollsoft.Rectangle attemptMatchOnNextPrevDisplay -int 1`

(restart the app after executing)

This will only take into account windows placed by Rectangle, and not windows placed manually, which is why I've never exposed this or made it a default.

**Comment by tahmed42** [comment]
Thanks a lot for the quick response @rxhanson. I set the config as you mentioned and appreciate that it works well for windows already placed by Rectangle. I haven't had the chance to reply more deeply until now, but wanted to continue the discussion.

I respect your push towards simplicity and the reality that displays are not all identical sizes. In my experience with Rectangle so far after using the attemptMatchOnNextPrevDisplay config, the most unintuitive moment happens when I move a plain window that hasn't yet been manipulated by Rectangle onto the next display. I'd expect it to appear in the "same" location on the next display (which happens to be an identically laid out display in my case), but instead it shifts to the center of the display, which feels jarring to me. Since everything gets moved to the center, moving multiple windows can cause a logjam in the center as well, which then necessitates rearranging when perhaps their placement on the prior display was already ideal or sufficient.

In other similar software, the approach has been to take the relative percentages of the current display's width and height that a window occupies and apply those to the new display. For example, a window that currently occupies the 7th through 34th percentiles (so to speak) of the current display's width and 45th through 68th percentiles of the current display's height, then map to those same percentiles in the new display even if the new display's resolution is different.

In the case of identical displays this maintains positioning exactly. In the case of shifting to a new display with the same aspect ratio (e.g going from 1920x1080 to 2560x1440) the positioning intuitively matches as well. In a more dramatic shift, say going from a landscape display to a portrait display, the relative space the window occupied is still matched (e.g. going from "somewhere around the top left corner" to a similarly matching "somewhere around the top left corner" on the new display) - which is behavior that a user can quickly learn and rely upon.

Having said that, the approach above absolutely does rely on assumptions on size/placement in an attempt to match a window's prior location. I definitely grant that there's other situations to account for as well, and I appreciate that Rectangle has often had intuitive fallback behavior in general (for example when dealing with a window that disallows resizing).

Just providing my intuition, and I acknowledge that it may not match Rectangle's philosophy. Either way, thanks a lot for creating Rectangle and for continuing to maintain it!

**Comment by rxhanson** [comment]
Thanks for the detailed follow up. I'm not set on keeping Rectangle the way it is forever, and would prefer to dynamically handle the window placement in a more desirable way (you're not the first one making this request). I just haven't had the time to add this in, and there are some gotchas, like handling gaps properly, that make this a decent time commitment. A good pull request for this is definitely welcome.
